### PR TITLE
Support percentage threshold for eviction 

### PR DESCRIPTION
**Pull request by ronnielai**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30400)

### CODE REVIEW

File: pkg/kubelet/eviction/eviction_manager.go [hunk 1]
```diff
@@ -62,6 +63,8 @@ type managerImpl struct {
 	resourceToRankFunc map[api.ResourceName]rankFunc
 	// resourceToNodeReclaimFuncs maps a resource to an ordered list of functions that know how to reclaim that resource.
 	resourceToNodeReclaimFuncs map[api.ResourceName]nodeReclaimFuncs
+	// machine info regarding the node
+	machineInfo *cadvisorapi.MachineInfo
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager.go]: isnt capacity available for each resource in summary provider?

**Comment by derekwaynecarr** [comment]
I'll take a deeper look tomorrow ;-)

**Comment by ronnielai** [comment]
I am still developing. Will ping you once I am done. Thx. :)

File: pkg/kubelet/eviction/eviction_manager.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager.go]: Unfortunately memory capacity is not available in summary.

**Comment by ronnielai** [comment]
PTAL

File: pkg/kubelet/eviction/eviction_manager.go (hunk 1 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager.go]: hmm... i would think to be consistent with `FsStats` we would add `Capacity` to `MemoryStats`, is there an objection to adding it?

File: pkg/kubelet/eviction/helpers.go [hunk 2]
```diff
@@ -329,7 +359,15 @@ func podMemoryUsage(podStats statsapi.PodStats) (api.ResourceList, error) {
 
 // formatThreshold formats a threshold for logging.
 func formatThreshold(threshold Threshold) string {
-	return fmt.Sprintf("threshold(signal=%v, operator=%v, value=%v, gracePeriod=%v)", threshold.Signal, threshold.Value.String(), threshold.Operator, threshold.GracePeriod)
+	return fmt.Sprintf("threshold(signal=%v, operator=%v, value=%v, gracePeriod=%v)", threshold.Signal, formatThresholdValue(threshold.Value), threshold.Operator, threshold.GracePeriod)
+}
+
+// formatThreshold formats a thresholdValue for logging.
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: // formatThresholdValue

File: pkg/kubelet/eviction/types.go [hunk 3]
```diff
@@ -66,14 +66,23 @@ type Config struct {
 	Thresholds []Threshold
 }
 
+type ThresholdValue struct {
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/types.go]: nit: some godoc like // ThresholdValue is a value holder that abstracts literal versus percentage based quantity

File: pkg/kubelet/eviction/types.go [hunk 4]
```diff
@@ -122,8 +131,16 @@ type statsFunc func(pod *api.Pod) (statsapi.PodStats, bool)
 // rankFunc sorts the pods in eviction order
 type rankFunc func(pods []*api.Pod, stats statsFunc)
 
+// signalObservation is the observed resoruce usage
+type singleObservation struct {
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/types.go]: s/singleObservation/signalObservation

Comment (derekwaynecarr) [inline pkg/kubelet/eviction/types.go]: s/resoruce/resource

File: pkg/kubelet/eviction/types.go [hunk 5]
```diff
@@ -122,8 +131,16 @@ type statsFunc func(pod *api.Pod) (statsapi.PodStats, bool)
...
+	// The resource apacity
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/types.go]: "capacity"

File: pkg/kubelet/eviction/eviction_manager.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager.go]: I think adding the field should be okay. I'll put in a TODO here.

File: pkg/kubelet/eviction/helpers.go [hunk 6]
```diff
@@ -532,15 +570,24 @@ func makeSignalObservations(summaryProvider stats.SummaryProvider) (signalObserv
 	// build an evaluation context for current eviction signals
 	result := signalObservations{}
 
-	if memory := summary.Node.Memory; memory != nil && memory.AvailableBytes != nil {
-		result[SignalMemoryAvailable] = resource.NewQuantity(int64(*memory.AvailableBytes), resource.BinarySI)
+	if memory := summary.Node.Memory; memory != nil && memory.AvailableBytes != nil && machineInfo != nil {
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: open question for all of these, but should we only make an observation if capacity is !=0 for each of these resources?  i am worried about subtle bugs in cadvisor potentially causing disastrous outcomes.

Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: i guess in this case, maybe just check if machineInfo != nil && machineInfo.MemoryCapacity != 0 ?

i may be worrying too much.

File: pkg/kubelet/eviction/helpers.go [hunk 7]
```diff
@@ -169,18 +170,47 @@ func parseThresholdStatement(statement string) (Threshold, error) {
...
-	quantity, err := resource.ParseQuantity(parts[1])
-	if err != nil {
-		return Threshold{}, err
+	quantityValue := parts[1]
+	if strings.HasSuffix(quantityValue, "%") {
+		percentage, err := parsePercentage(quantityValue)
+		if err != nil {
+			return Threshold{}, err
+		}
+		if percentage < 0 {
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: shouldn't this be <= 0?

basically, a percentage must have a positive value?

is there a case where total exhaustion is a good thing?

maybe this also applies to our current literal value support.

File: pkg/kubelet/eviction/helpers.go (hunk 6 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: If capacity is 0, the threshold will be set to 0 (0 * percentage), which kinda behaves like a fail-safe. Did I misunderstand your point?

File: pkg/kubelet/eviction/helpers.go (hunk 7 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Yeah I was following the existing code :)

I'll changed to <= for both cases.

File: pkg/kubelet/eviction/eviction_manager.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager.go]: From the comment it seems that I can get the limit from AvailableBytes + WorkingSetBytes. I'll see how that goes.

File: pkg/kubelet/eviction/helpers.go (hunk 6 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: I'll see if I can get rid of machineInfo all together.

File: pkg/kubelet/eviction/helpers.go (hunk 2 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Done

File: pkg/kubelet/eviction/types.go (hunk 3 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/types.go]: Don

File: pkg/kubelet/eviction/types.go (hunk 4 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/types.go]: Done

File: pkg/kubelet/eviction/types.go (hunk 5 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/types.go]: Done

**Comment by ronnielai** [comment]
Addressed comments. PTAL.

File: pkg/kubelet/eviction/helpers.go [hunk 8]
```diff
@@ -532,15 +569,24 @@ func makeSignalObservations(summaryProvider stats.SummaryProvider) (signalObserv
 	// build an evaluation context for current eviction signals
 	result := signalObservations{}
 
-	if memory := summary.Node.Memory; memory != nil && memory.AvailableBytes != nil {
-		result[SignalMemoryAvailable] = resource.NewQuantity(int64(*memory.AvailableBytes), resource.BinarySI)
+	if memory := summary.Node.Memory; memory != nil && memory.AvailableBytes != nil && memory.WorkingSetBytes != nil {
+		result[SignalMemoryAvailable] = signalObservation{
+			available: resource.NewQuantity(int64(*memory.AvailableBytes), resource.BinarySI),
+			capacity:  resource.NewQuantity(int64(*memory.AvailableBytes+*memory.WorkingSetBytes), resource.BinarySI),
```
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: +1

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by ronnielai** [comment]
Mark P2 because there are several PRs affected by this.

**Pull request merged**